omega/omega-context/src/main/java/io/servicecomb/saga/omega/context/OmegaContext.java

```diff
@@ -73,26 +78,27 @@ public void setParentTxId(String parentTxId) {
   }
 
   // TODO: 2017/12/23 remove this context entry by the end of its corresponding global tx
-  public void addContext(String id, Object target, String compensationMethod, Object... args) {
-    compensationContexts.put(id, new CompensationContext(target, compensationMethod, args));
+  public void addContext(String globalTxId, String localTxId, Object target, String compensationMethod, Object... args) {
+    compensationContexts.computeIfAbsent(globalTxId, k -> new ConcurrentHashMap<>())
+        .put(localTxId, new CompensationContext(target, compensationMethod, args));
   }
 
   public boolean containsContext(String globalTxId) {
     return compensationContexts.containsKey(globalTxId);
   }
 
   public void compensate(String globalTxId) {
-    CompensationContext compensationContext = compensationContexts.get(globalTxId);
-
-    try {
-      invokeMethod(compensationContext);
-    } catch (NoSuchMethodException | IllegalAccessException | InvocationTargetException e) {
-      throw new IllegalStateException(
-          "Pre-checking for compensate method " + compensationContext.compensationMethod
-              + " was somehow skipped, did you forget to configure compensable method checking on service startup?",
-          e);
-    } finally {
-      compensationContexts.remove(globalTxId);
+    Map<String, CompensationContext> contexts = compensationContexts.remove(globalTxId);
```

Comment:
If the process is crashed this time,  we will lose all the compensation contexts.

Reply:
not really, it's sent to alpha. we just didn't wire it up yet.